daemon/image.go

```diff
@@ -113,12 +121,228 @@ func (daemon *Daemon) GetCachedImage(imgID image.ID, config *containertypes.Conf
+// Define a struct to store this mapping for convenience.
+type historyWithSourceT struct {
+	// sourceLayerID is the layer on which the command was run. Empty digest if this is the first command or
+	// if nothing has actually been added to the rootfs yet.
+	sourceLayerID layer.DiffID
+	// cmd is the command which got run on sourceLayerID
+	cmd string
+	// resulingLayerID is the result of running cmd on sourceLayerID (might be the same as sourceLayerID)
+	resultingLayerID layer.DiffID
+	// createdAt is the time the history entry was created
+	createdAt time.Time
+	// history is the actual, underlying history entry
+	history image.History
+}
+
+func makeHistoryWithSource(image *image.Image) []historyWithSourceT {
+	// Let's make those structs now
+	historyWithSource := make([]historyWithSourceT, len(image.History))
+	layerIndex := -1
+	for i, h := range image.History {
+
+		// previous is layerIndex from previous iteration
+		if layerIndex == -1 {
+			historyWithSource[i].sourceLayerID = digest.DigestSha256EmptyTar
+		} else {
+			historyWithSource[i].sourceLayerID = image.RootFS.DiffIDs[layerIndex]
+		}
+
+		// now increment, if needed, and look at the result layer ID
+		if !h.EmptyLayer {
+			layerIndex = layerIndex + 1
+		}
+		if layerIndex == -1 {
+			historyWithSource[i].resultingLayerID = digest.DigestSha256EmptyTar
+		} else {
+			historyWithSource[i].resultingLayerID = image.RootFS.DiffIDs[layerIndex]
+		}
+
+		// Copy the other history entries over I'm interested in
+		historyWithSource[i].cmd = h.CreatedBy
+		historyWithSource[i].createdAt = h.Created
+		historyWithSource[i].history = h
+	}
+
+	return historyWithSource
+}
+
 // GetCachedImageOnBuild returns a reference to a cached image whose parent equals `parent`
 // and runconfig equals `cfg`. A cache miss is expected to return an empty ID and a nil error.
-func (daemon *Daemon) GetCachedImageOnBuild(imgID string, cfg *containertypes.Config) (string, error) {
-	cache, err := daemon.GetCachedImage(image.ID(imgID), cfg)
-	if cache == nil || err != nil {
+func (cache *daemonImageCacheForBuild) GetCachedImageOnBuild(imgID string, cfg *containertypes.Config) (string, error) {
+	cachedImage, err := cache.daemon.GetCachedImage(image.ID(imgID), cfg)
+	if err != nil {
 		return "", err
 	}
-	return cache.ID().String(), nil
+	if cachedImage != nil {
+		// We found a cache hit using the old parent image method
+		return cachedImage.ID().String(), nil
+	}
+
+	// We didn't find a cache hit using that method. Explore cacheFrom images for matching history
+	var parentImage *image.Image
+	var parentImageHistory []historyWithSourceT
+	// i.e. not FROM SCRATCH
+	if imgID != "" {
+		parentImage, err = cache.daemon.imageStore.Get(image.ID(imgID))
+		if err != nil {
+			return "", err
+		}
+		parentImageHistory = makeHistoryWithSource(parentImage)
+	}
+
+	// For each thing we are caching from, see if it matches parentImageHistory
+	type matchStruct struct {
```

Comment:
Likewise, I can probably find a way to get rid of this type and rephrase the code in terms of existing types.